clang/lib/Parse/ParseDecl.cpp

```diff
@@ -3703,8 +3703,14 @@ void Parser::ParseDeclarationSpecifiers(
           // We reject AT_LifetimeBound and AT_AnyX86NoCfCheck, even though they
           // are type attributes, because we historically haven't allowed these
           // to be used as type attributes in C++11 / C23 syntax.
-          if (PA.isTypeAttr() && PA.getKind() != ParsedAttr::AT_LifetimeBound &&
-              PA.getKind() != ParsedAttr::AT_AnyX86NoCfCheck)
+          if (PA.getKind() == ParsedAttr::AT_LifetimeBound) {
+            Diag(PA.getLoc(), diag::err_attribute_wrong_decl_type_str)
+                << PA << PA.isRegularKeywordAttribute()
+                << "parameters and implicit object parameters";
```

Comment:
Why are we not using %select for that?

Beside, is "implicit object parameters" sufficiently understandable? - or should we say "member functions"?

Reply:
> Why are we not using %select for that?

Can you explain a bit more? IIUC, you're suggesting using the `err_attribute_wrong_decl_type` the one with the "%select"? That is possible, but then we'll need to add a new Kind in `AttributeDeclKind`.

> Beside, is "implicit object parameters" sufficiently understandable? - or should we say "member functions"?

I think the implicit object parameter is fine. "member functions" is not entirely correct, the attribute is on the implicit object parameter even though it is applied to a function type in the implementation.

Reply:
I would like to second @cor3ntin's comment. We should aim to have all message strings in one place (i.e. the .td file), it makes searching for error messages much easier.

However, `err_attribute_wrong_decl_type_str` in particular seems to be using strings in the source code ([1](https://github.com/llvm/llvm-project/blob/0cda970ecc8a885acf7298a61370a1368b0ea39b/clang/lib/Sema/SemaDeclAttr.cpp#L5529), [2](https://github.com/llvm/llvm-project/blob/0cda970ecc8a885acf7298a61370a1368b0ea39b/clang/lib/Sema/SemaDeclAttr.cpp#L1842), [3](https://github.com/llvm/llvm-project/blob/0cda970ecc8a885acf7298a61370a1368b0ea39b/clang/lib/Sema/SemaDeclAttr.cpp#L2736)).
So I think this change is consistent with that, and I think we should just land it and replace this with a follow-up instead (@emaxx-google would you be willing to do it?)

Adding another `AttributeDeclKind` seems off, given that it's a enumeration used primarily by parser. But adding a new enumeration for select (that maps to `AttributeDeclKind` for the known values and adds a few more) seems like a good path forward.

Reply:
I'm happy to follow up with the change!

Reply:
I do see that `err_attribute_wrong_decl_type_str` uses that same method in other places which is unfortunate. It makes looking for diagnostics a lot harder.